codec-dns/src/main/java/io/netty/handler/codec/dns/DefaultDnsRecordEncoder.java

```diff
@@ -49,29 +54,93 @@ public void encodeRecord(DnsRecord record, ByteBuf out) throws Exception {
             encodeQuestion((DnsQuestion) record, out);
         } else if (record instanceof DnsPtrRecord) {
             encodePtrRecord((DnsPtrRecord) record, out);
+        } else if (record instanceof DnsOptEcsRecord) {
+            encodeOptEcsRecord((DnsOptEcsRecord) record, out);
+        } else if (record instanceof DnsOptPseudoRecord) {
+            encodeOptPseudoRecord((DnsOptPseudoRecord) record, out);
         } else if (record instanceof DnsRawRecord) {
             encodeRawRecord((DnsRawRecord) record, out);
         } else {
             throw new UnsupportedMessageTypeException(StringUtil.simpleClassName(record));
         }
     }
 
-    private void encodePtrRecord(DnsPtrRecord record, ByteBuf out) throws Exception {
+    private void encodeRecord0(DnsRecord record, ByteBuf out) throws Exception {
         encodeName(record.name(), out);
-
         out.writeShort(record.type().intValue());
         out.writeShort(record.dnsClass());
         out.writeInt((int) record.timeToLive());
+    }
 
+    private void encodePtrRecord(DnsPtrRecord record, ByteBuf out) throws Exception {
+        encodeRecord0(record, out);
         encodeName(record.hostname(), out);
     }
 
-    private void encodeRawRecord(DnsRawRecord record, ByteBuf out) throws Exception {
-        encodeName(record.name(), out);
+    private void encodeOptPseudoRecord(DnsOptPseudoRecord record, ByteBuf out) throws Exception {
+        encodeRecord0(record, out);
+        out.writeShort(0);
+    }
 
-        out.writeShort(record.type().intValue());
-        out.writeShort(record.dnsClass());
-        out.writeInt((int) record.timeToLive());
+    private void encodeOptEcsRecord(DnsOptEcsRecord record, ByteBuf out) throws Exception {
+        encodeRecord0(record, out);
+
+        int sourcePrefixLength = record.sourcePrefixLength();
+        int scopePrefixLength = record.scopePrefixLength();
+        int leftOverBits = sourcePrefixLength & PREFIX_MASK;
+        InetAddress address = record.address();
+
+        byte[] bytes = address.getAddress();
+        int addressBits = bytes.length * Byte.SIZE;
+        if (addressBits < sourcePrefixLength || sourcePrefixLength <= 0) {
+            throw new IllegalArgumentException(sourcePrefixLength + ": "
+                    + sourcePrefixLength + " (expected: 0 >= " + addressBits + ')');
+        }
+
+        // See http://www.iana.org/assignments/address-family-numbers/address-family-numbers.xhtml
+        final short addressNumber;
+        if (address instanceof Inet4Address) {
+            addressNumber = 1;
+        } else if (address instanceof Inet6Address) {
+            addressNumber = 2;
+        } else {
+            throw new IllegalArgumentException("address " + address + " not supported");
+        }
+
+        // Guard aginst value <= Byte.SIZE
+        int payloadLength = Math.max(1, (int) Math.rint((double) sourcePrefixLength / Byte.SIZE));
```

Comment:
A few questions...

1. Why does this have a max value of 1? Isn't it possible that the address may include more than 1 byte?
2. should we allow for the address to consume 0 bytes? The RFC isn't overly clear about this case but it seems like `SOURCE PREFIX-LENGTH` of `0` means we can just omit the address.

https://tools.ietf.org/html/rfc7871#section-7.1.2:
> If there is no ADDRESS set, i.e.,
   SOURCE PREFIX-LENGTH is set to 0, then FAMILY SHOULD be set to the
   transport over which the query is sent.

https://tools.ietf.org/html/rfc7871#section-7.5
> If an Intermediate Nameserver receives a query with SOURCE PREFIX-
   LENGTH set to 0, it MUST NOT include client address information in
   queries made to resolve that client's request (see Section 7.1.2).